src/combinatorics/young_diagram.lean

```diff
@@ -123,4 +128,148 @@ end distrib_lattice
 /-- Cardinality of a Young diagram -/
 @[reducible] protected def card (μ : young_diagram) : ℕ := μ.cells.card
 
+section transpose
+/-- The `transpose` of a Young diagram is obtained by swapping i's with j's. --/
+
+def transpose (μ : young_diagram) : young_diagram :=
+{ cells :=  (equiv.prod_comm _ _).finset_congr μ.cells,
+  is_lower_set := λ _ _ h, begin
+    simp only [finset.mem_coe, equiv.finset_congr_apply, finset.mem_map_equiv],
+    intro hcell,
+    apply μ.is_lower_set _ hcell,
+    simp [h],
+  end }
+
+@[simp] lemma mem_transpose {μ : young_diagram} {c : ℕ × ℕ} : c ∈ μ.transpose ↔ c.swap ∈ μ :=
+by { change c ∈ μ.cells.map _ ↔ _, rw finset.mem_map_equiv, refl, }
+
+@[simp] lemma transpose_transpose (μ : young_diagram) : μ.transpose.transpose = μ :=
+by { ext, simp }
+
+lemma transpose_eq_iff_eq_transpose {μ ν : young_diagram} :
+  μ.transpose = ν ↔ μ = ν.transpose :=
+by { split; { rintro rfl, simp } }
+
+@[simp] lemma transpose_eq_iff {μ ν : young_diagram} :
+  μ.transpose = ν.transpose ↔ μ = ν :=
+by { rw transpose_eq_iff_eq_transpose, simp }
+
+-- This is effectively both directions of the iff statement below.
+protected lemma le_of_transpose_le {μ ν : young_diagram} (h_le : μ.transpose ≤ ν) :
+  μ ≤ ν.transpose :=
+λ c hc, by { simp only [mem_transpose], apply h_le, simpa }
+
+@[simp] lemma transpose_le_iff {μ ν : young_diagram} : μ.transpose ≤ ν.transpose ↔ μ ≤ ν :=
+⟨ λ h, by { convert young_diagram.le_of_transpose_le h, simp },
+  λ h, by { convert @young_diagram.le_of_transpose_le _ _ _, simpa } ⟩
+
+@[mono]
+protected lemma transpose_mono {μ ν : young_diagram} (h_le : μ ≤ ν) : μ.transpose ≤ ν.transpose :=
+by simp [h_le]
+
+/-- Transposing Young diagrams is an `order_iso`. -/
+def transpose_order_iso : young_diagram ≃o young_diagram :=
+⟨⟨transpose, transpose, λ _, by simp, λ _, by simp⟩, by simp⟩
+
+end transpose
+
+section rows
+/-! ### Rows and row lengths of Young diagrams.
+
+This section defines `μ.row` and `μ.row_len`, with the following API:
+      1.  `(i, j) ∈ μ ↔ j < μ.row_len i`
+      2.  `μ.row i = {i} ×ˢ (finset.range (μ.row_len i))`
+      3.  `μ.row_len i = (μ.row i).card`
+      4.  `∀ {i1 i2}, i1 ≤ i2 → μ.row_len i2 ≤ μ.row_len i1`
+
+Note: #3 is not convenient for defining `μ.row_len`; instead, `μ.row_len` is defined
+as the smallest `j` such that `(i, j) ∉ μ`. --/
+
+/-- The `i`-th row of a Young diagram consists of the cells whose first coordinate is `i`. --/
+def row (μ : young_diagram) (i : ℕ) : finset (ℕ × ℕ) := μ.cells.filter (λ c, c.fst = i)
+
+lemma mem_row_iff {μ : young_diagram} {i : ℕ} {c : ℕ × ℕ} : c ∈ μ.row i ↔ c ∈ μ ∧ c.fst = i :=
+by simp [row]
```

Comment:
It seems like it would be useful to have this version too:
```lean
lemma mk_mem_row_iff {μ : young_diagram} {i j : ℕ} : (i, j) ∈ μ.row i ↔ (i, j) ∈ μ :=
by simp [row]
```
This eliminates the additional `=` on the RHS.